text/0023-entity-evaluated-attributes.md

```diff
@@ -0,0 +1,92 @@
+are stored as JSON data, so it is easiest to parse entity attributes into `RestrictedExpression`'s, rather than launching an evaluator at parse time. However, with a 
+more generalized storage format, it is likely that this will become a stumbling block. For example, attribute values might best be stored in some native format (e.g.
+integers stored as 64-bit integers, decimals stored as native floats rather than a call to the extension function `decimal`), and it is silly to try to convert these first to
+a `RestrictedExpression` that is reevaluated, rather than just converting it directly to a `PartialValue`.
+
+Worse, if we do not want to load all of the entities into memory at once (as in [RFC 16](https://github.com/cedar-policy/rfcs/pull/16)), we currently have no way to support this because our evaluator goes through and evaluates all the
+entity attributes up front. We could solve this through lazy attribute caching, as the other RFC suggests, but that has several drawbacks, which the other RFC details.
+With this RFC, lazy caching becomes unnecessary because the user could parse the attributes up front themselves (or load them from some native format).
+
+## Detailed design
+
+The easiest way to make this change internally is to make `Entity` and `Entities` generic for some `Entity<T>`, where `T` is the attribute type, and similarly, we change
+`Entities` to `Entities<T>`.
+
+Externally, to leave the API unchanged, we can make separate wrappers `Entity` and `ParsedEntity` around `Entity<RestrictedExpression>` and `Entity<PartialValue>` respectively.
+
+In order to leave the existing API unchanged, we can leave the current methods `is_authorized` which simply call `eval_attrs()` themselves, adding a new method,
+`is_authorized_parsed` (TODO: figure out name) which accepts `ParsedEntities` instead of `Entities`.
+
+## Drawbacks
+
+- It's a relatively large refactor because it touches fundamental Cedar types (`Entity` and `Entities`). It's actually harder than simply adding a lazy attribute cacher while offering
+fewer features (but in some sense allowing more flexibility, since the user can do the caching themselves using `RefCell<T>`/`unsafe` if they really want, or choose not
+to do caching at all).
+- It adds new API very similar to existing API, which could be confusing.
+- It may require us to expose/create `PartialValue` (hence `Expr`) API which is more stuff to maintain
+
+## Alternatives
+
+- Get rid of attribute caching entirely and re-evaluate the attribute every time the attribute is needed. Honestly, this isn't even that bad, since restricted expression evaluation
+is linear time, essentially doing a `clone()` operation, right? We routinely do `clone`'s on `PartialValue`'s in the code.
+- Use lazy attribute caching as described in the [other RFC](https://github.com/cedar-policy/rfcs/pull/16)
```

Comment:
If we're going to want this eventually anyways, would it make sense to change `Entity` to `MaybeEvaledEntity` and support `EvaledEntity`, `UnevaledEntity` and mixtures thereof?

Reply:
I'm not sure lazy attribute caching is on the roadmap for Cedar itself. Instead, other extensions to Cedar can implement their own entity stores (which implement the `EntityDataSource` trait) and do caching themselves, and they can do this caching however they want. Ultimately, they will be returning `Value`s, but they may decide internally to store restricted expressions.